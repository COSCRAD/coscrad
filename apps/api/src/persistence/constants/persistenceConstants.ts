/**
 * **Caution**
 * This prefix is used in auto-generated databases for test. It should never be
 * included anywhere in a production database (or any database you wish to persist).
 * Databases whose names include this string anywhere in their name will be
 * automatically dropped during test cleanup.
 */
export const TEST_DATABASE_PREFIX = 'testonly';

export const TEST_DATABASE_NAME_LENGTH = 20;

// dependency injection token
export const REPOSITORY_PROVIDER = 'REPOSITORY_PROVIDER';
