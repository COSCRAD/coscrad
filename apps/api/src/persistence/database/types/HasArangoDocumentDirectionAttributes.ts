import { ArangoDocumentHandle } from './ArangoDocumentHandle';

type ToAndFrom = {
    _to: ArangoDocumentHandle;

    _from: ArangoDocumentHandle;
};

export type HasArangoDocumentDirectionAttributes<T = undefined> = T extends undefined
    ? ToAndFrom
    : ToAndFrom & T;
