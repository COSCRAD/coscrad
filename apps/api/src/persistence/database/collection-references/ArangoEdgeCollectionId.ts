import { ValueType } from '../../../lib/types/valueType';

export const ArangoEdgeCollectionId = {
    categoryEdgeCollectionID: 'category_edges',
    edgeConnectionCollectionID: 'resource_edge_connections',
} as const;

export type ArangoEdgeCollectionId = ValueType<typeof ArangoEdgeCollectionId>;

export const isArangoEdgeCollectionCollectionID = (
    input: unknown
): input is ArangoEdgeCollectionId =>
    Object.values(ArangoEdgeCollectionId).includes(input as ArangoEdgeCollectionId);

export const getAllArangoEdgeCollectionIDs = (): ArangoEdgeCollectionId[] =>
    Object.values(ArangoEdgeCollectionId);
