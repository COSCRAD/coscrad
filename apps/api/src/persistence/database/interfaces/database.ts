import { ResourceId } from 'apps/api/src/domain/types/ResourceId';
import { Maybe } from 'apps/api/src/lib/types/maybe';

export interface IDatabase {
    fetchById: <TEntityDTO>(id: ResourceId, collectionName: string) => Promise<Maybe<TEntityDTO>>;

    fetchMany: <TEntityDTO>(collectionName: string) => Promise<Maybe<TEntityDTO[]>>;

    create: <TEntityDTO>(dto: TEntityDTO, collectionName: string) => Promise<void>;

    createMany: <TEntityDTO>(dto: TEntityDTO[], collectionName: string) => Promise<void>;

    getCount: (collectionName: string) => Promise<number>;

    update: <TEntityDTO>(id: ResourceId, dto: TEntityDTO, collectionName: string) => Promise<void>;
}
