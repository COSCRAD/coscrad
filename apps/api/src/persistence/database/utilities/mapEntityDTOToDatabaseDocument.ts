import { EdgeConnection } from '../../../domain/models/context/edge-connection.entity';
import { isAggregateId } from '../../../domain/types/AggregateId';
import { HasAggregateId } from '../../../domain/types/HasAggregateId';
import { DTO } from '../../../types/DTO';
import { HasArangoDocumentDirectionAttributes } from '../types/HasArangoDocumentDirectionAttributes';

// TODO Rename this, the common base type for edge and non-edge documents in arango
export type ArangoDocumentForAggregateRoot<TEntityDTO extends HasAggregateId = HasAggregateId> =
    Omit<TEntityDTO, 'id'> & {
        _key: string;
    };

export type ArangoDatabaseDocument<TEntity extends HasAggregateId> = TEntity extends EdgeConnection
    ? HasArangoDocumentDirectionAttributes<ArangoDocumentForAggregateRoot<TEntity>>
    : ArangoDocumentForAggregateRoot<TEntity>;

export default <T extends HasAggregateId>(entityDTO: DTO<T>): ArangoDocumentForAggregateRoot<T> =>
    Object.entries(entityDTO).reduce(
        (accumulatedMappedObject: ArangoDocumentForAggregateRoot<T>, [key, value]) => {
            if (key === 'id') {
                // Note invalid ids will be omitted from the output
                if (isAggregateId(value)) accumulatedMappedObject['_key'] = value as string;
            } else {
                accumulatedMappedObject[key] = value;
            }

            return accumulatedMappedObject as unknown as ArangoDocumentForAggregateRoot<T>;
        },
        {} as ArangoDocumentForAggregateRoot<T>
    );
