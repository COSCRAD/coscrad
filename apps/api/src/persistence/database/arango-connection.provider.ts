import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Database } from 'arangojs';
import { Scheme } from '../../app/config/constants/Scheme';
import { isNullOrUndefined } from '../../domain/utilities/validation/is-null-or-undefined';
import { DTO } from '../../types/DTO';
import ArangoDatabaseConfiguration from './ArangoDatabaseConfiguration';
import { getAllArangoCollectionIDs } from './collection-references/ArangoCollectionId';
import { isArangoEdgeCollectionCollectionID } from './collection-references/ArangoEdgeCollectionId';
import DatabaseAlreadyInitializedError from './errors/DatabaseAlreadyInitializedError';
import DatabaseCannotBeDroppedError from './errors/DatabaseCannotBeDroppedError';
import DatabaseNotYetInitializedError from './errors/DatabaseNotYetInitializedError';
import canDatabaseBeDropped from './utilities/canDatabaseBeDropped';

// Alias for more clarity from the outside; TODO wrap `Database` with simpler API?
export type ArangoConnection = Database;

type Port = `${number}`;

const isPortRequired = (scheme: Scheme, port: `${number}`) => {
    if (scheme === 'http' && port === '80') return false;
    if (scheme === 'https' && port === '443') return false;
    return true;
};

const buildFullHostURL = (
    domain: string,
    scheme: Scheme = Scheme.https,
    port: Port = '443'
): string => `${scheme}://${domain}${isPortRequired(scheme, port) ? `:${port}` : ''}`;

type AdminDatabaseConnectionOptions = {
    shouldConnectToSystemDatabase: boolean;
};

@Injectable()
export class ArangoConnectionProvider {
    #connection: ArangoConnection;

    #databaseConfiguration: ArangoDatabaseConfiguration;

    isInitialized = false;

    public get databaseConfiguration() {
        return this.#databaseConfiguration.getConfig();
    }

    constructor(private configService: ConfigService) {
        this.setDatabaseConfiguration({
            dbName: this.configService.get<string>('ARANGO_DB_NAME'),
            dbRootPass: this.configService.get<string>('ARANGO_DB_ROOT_PASSWORD'),
            dbUser: this.configService.get<string>('ARANGO_DB_USER'),
            dbPass: this.configService.get<string>('ARANGO_DB_USER_PASSWORD'),
            dbHostUrl: buildFullHostURL(
                this.configService.get<string>('ARANGO_DB_HOST_DOMAIN', 'localhost'),
                this.configService.get<Scheme>('ARANGO_DB_HOST_SCHEME', Scheme.http),
                this.configService.get<number>('ARANGO_DB_HOST_PORT', 80).toString() as Port
            ),
        });

        console.log(`dbName: ${this.databaseConfiguration.dbName}, dbRootPass: ${this.databaseConfiguration.dbRootPass}, dbUser: ${this.databaseConfiguration.dbUser}, dbPass: ${this.databaseConfiguration.dbPass}, dbHostUrl: ${this.databaseConfiguration.dbHostUrl}`);
        

        const systemDB = this.#getAdminDBConnection({
            shouldConnectToSystemDatabase: true,
        });

        // TODO why is `useDatabase` deprecated? can we use myDB.database("db_name")?
        const dbInstance = systemDB.database(this.databaseConfiguration.dbName);
        dbInstance.useBasicAuth(
            this.databaseConfiguration.dbUser,
            this.databaseConfiguration.dbPass
        );

        this.#connection = dbInstance;
    }

    /**
     * This async setup helper will
     * - Create the database (the one referenced in your .env via ConfigService)
     * only if this database does not exist
     * - Add any missing collections (but leave existing collections untouched)
     * and can only be run once.
     */
    async initialize(): Promise<void> {
        if (this.isInitialized) throw new DatabaseAlreadyInitializedError();

        await this.#createDatabaseIfNotExists();

        await this.#createAllMissingCollections();

        this.isInitialized = true;
    }

    getDatabaseName(): string {
        return this.databaseConfiguration.dbName;
    }

    getConnection(): ArangoConnection {
        if (!this.isInitialized)
            throw new DatabaseNotYetInitializedError('get an Arango connection');

        return this.#connection;
    }

    public setDatabaseConfiguration(config: DTO<ArangoDatabaseConfiguration>) {
        this.#databaseConfiguration = new ArangoDatabaseConfiguration({
            ...config,
        });
    }

    #getAdminDBConnection({
        shouldConnectToSystemDatabase,
    }: AdminDatabaseConnectionOptions): Database {
        const adminInstance = new Database({
            url: this.databaseConfiguration.dbHostUrl,
        });

        adminInstance.useBasicAuth('root', this.databaseConfiguration.dbRootPass);

        if (shouldConnectToSystemDatabase) {
            return adminInstance;
        }

        // TODO why is `useDatabase` deprecated? can we use myDB.database("db_name")?
        const dbInstance = adminInstance.database(this.databaseConfiguration.dbName);

        return dbInstance;
    }

    async #doesCollectionExist(collectionName: string): Promise<boolean> {
        const allCollections = await this.#connection
            .listCollections()
            .then((allCollectionMetadata) => allCollectionMetadata.map(({ name }) => name));

        const doesCollectionExist = allCollections.includes(collectionName);

        return doesCollectionExist;
    }

    async #createDatabaseIfNotExists(): Promise<void> {
        const databaseName = this.databaseConfiguration.dbName;

        if (await this.#doesDatabaseExist(databaseName)) return;

        const { dbUser } = this.#databaseConfiguration;

        await this.#getAdminDBConnection({
            shouldConnectToSystemDatabase: true,
        }).createDatabase(databaseName, {
            users: [{ username: dbUser }, { username: 'root' }],
        });
    }

    async #createAllMissingCollections(): Promise<void> {
        await Promise.all(
            getAllArangoCollectionIDs().map((collectionName) =>
                this.#createCollectionIfNotExists(collectionName)
            )
        );
    }

    async #createCollectionIfNotExists(collectionName: string): Promise<void> {
        const doesCollectionExist = await this.#doesCollectionExist(collectionName);

        if (doesCollectionExist) return;

        /**
         * TODO [https://www.pivotaltracker.com/story/show/182132515]
         * cleanup the references
         */
        if (isArangoEdgeCollectionCollectionID(collectionName)) {
            await this.#connection.createEdgeCollection(collectionName);
        } else {
            await this.#connection.createCollection(collectionName);
        }
    }

    #doesDatabaseExist = async (databaseName: string) => {
        if (isNullOrUndefined(databaseName)) return false;

        const doesDatabaseExist = await this.#getAdminDBConnection({
            shouldConnectToSystemDatabase: true,
        })
            .listDatabases()
            .then((allDatabaseNames) => allDatabaseNames.includes(databaseName));

        return doesDatabaseExist;
    };

    /**
     * **CAUTION** This method is only to be used in tests.
     */
    async dropDatabaseIfExists(): Promise<void> {
        const dbName = this.databaseConfiguration.dbName;

        const doesDatabaseExist = await this.#doesDatabaseExist(dbName);

        if (!doesDatabaseExist) return;

        // for safety, you can only drop a db whose name matches the test db name pattern
        if (!canDatabaseBeDropped(dbName)) throw new DatabaseCannotBeDroppedError(dbName);

        await this.#getAdminDBConnection({
            shouldConnectToSystemDatabase: true,
        }).dropDatabase(dbName);
    }
}
