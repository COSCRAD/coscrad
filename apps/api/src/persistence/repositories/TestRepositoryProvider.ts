import { Entity } from '../../domain/models/entity';
import { EntityType, InMemorySnapshot } from '../../domain/types/entityType';
import { DatabaseProvider } from '../database/database.provider';
import { getArangoCollectionIDFromEntityType } from '../database/get-arango-collection-ids';
import { RepositoryProvider } from './repository.provider';

export default class TestRepositoryProvider extends RepositoryProvider {
  constructor(databaseProvider: DatabaseProvider) {
    super(databaseProvider);
  }

  // TODO We should correlate entity type with TEntity here
  public async addEntitiesOfSingleType<TEntity extends Entity>(
    entityType: EntityType,
    entities: TEntity[]
  ): Promise<void> {
    await this.forEntity<TEntity>(entityType).createMany(entities);
  }

  // TODO fix types
  public async addEntitiesOfManyTypes(
    snapshot: InMemorySnapshot
  ): Promise<void> {
    const writePromises = Object.entries(snapshot).map(
      ([entityType, entityInstances]) =>
        this.addEntitiesOfSingleType(
          entityType as EntityType,
          entityInstances as any
        )
    );
  }

  public async deleteAllEntitiesOfGivenType(
    entityType: EntityType
  ): Promise<void> {
    await (
      await this.databaseProvider.getDBInstance()
    ).deleteAll(getArangoCollectionIDFromEntityType(entityType));
  }
}
