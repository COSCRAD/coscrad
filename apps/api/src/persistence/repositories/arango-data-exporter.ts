import { existsSync, mkdirSync, writeFileSync } from 'fs';
import { InternalError } from '../../lib/errors/InternalError';
import { isNotFound } from '../../lib/types/not-found';
import { InMemoryDatabaseSnapshot } from '../../test-data/utilities/convertInMemorySnapshotToDatabaseFormat';
import { ArangoQueryRunner } from '../database/arango-query-runner';

/**
 * TODO [test-coverage] [https://github.com/COSCRAD/coscrad/pull/381#discussion_r1198018120]
 * We need a test of this class.
 */
export class ArangoDataExporter {
    constructor(private readonly arangoQueryRunner: ArangoQueryRunner) {}

    async fetchSnapshot(knownEdgeCollections: string[] = []): Promise<InMemoryDatabaseSnapshot> {
        const document = {};

        const edge = {};

        const allCollections = await this.arangoQueryRunner.collections();

        const documentCollectionNames = allCollections
            .map(({ name }) => name)
            .filter((c) => !knownEdgeCollections.includes(c));

        /**
         * TODO Consider parallelizing these queries for performance.
         */
        for (const collection of documentCollectionNames) {
            const queryResult = await this.arangoQueryRunner.export(collection);

            if (isNotFound(queryResult)) {
                throw new InternalError(
                    `Failed to export. Collection: ${collection} does not exist`
                );
            }

            document[collection] = queryResult;
        }

        for (const collection of knownEdgeCollections) {
            // TODO handle the case that the collection does not exist explicitly
            const queryResult = await this.arangoQueryRunner.export(collection);

            if (isNotFound(queryResult)) {
                throw new InternalError(
                    `Failed to export. Collection: ${collection} does not exist`
                );
            }

            edge[collection] = queryResult;
        }

        return {
            document,
            edge,
        } as InMemoryDatabaseSnapshot;
    }

    async dumpSnapshot(directory: string, filename: string, knownEdgeCollections: string[] = []) {
        if (!existsSync(directory)) {
            mkdirSync(directory);
        }

        const snapshot = await this.fetchSnapshot(knownEdgeCollections);

        try {
            // TODO We need an abstraction for writing files
            writeFileSync(`${directory}/${filename}`, JSON.stringify(snapshot, null, 4));
        } catch (error) {
            throw new Error(`failed to write snapshot`);
        }
    }

    async restoreFromSnapshot(snapshot: InMemoryDatabaseSnapshot) {
        /**
         * This functions like the pop-out tab on a VHS or audio cassette in that
         * it enforces us to be intentional about writing data. By leaving this
         * env variable unset in production, we will avoid accidentally importing
         * data there.
         */
        if (process.env['DATA_MODE'] !== 'import') {
            throw new InternalError(`You must set $DATA_MODE=import to enable imports.`);
        }

        const { document: allDocumentSnapshots, edge: allEdgeSnapshots } = snapshot;

        for (const [collectionName, { checksum, documents }] of [
            ...Object.entries(allDocumentSnapshots),
        ]) {
            await this.arangoQueryRunner.import(collectionName, documents, 'document', checksum);
        }

        for (const [collectionName, { checksum, documents }] of [
            ...Object.entries(allEdgeSnapshots),
        ]) {
            await this.arangoQueryRunner.import(collectionName, documents, 'edge', checksum);
        }
    }
}
