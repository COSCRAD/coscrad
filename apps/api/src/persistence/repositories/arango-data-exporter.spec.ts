import { INestApplication } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { Test } from '@nestjs/testing';
import { existsSync, readFileSync, unlinkSync } from 'fs';
import buildMockConfigService from '../../app/config/__tests__/utilities/buildMockConfigService';
import buildConfigFilePath from '../../app/config/buildConfigFilePath';
import { Environment } from '../../app/config/constants/Environment';
import { CoscradEventFactory } from '../../domain/common';
import { DeluxeInMemoryStore } from '../../domain/types/DeluxeInMemoryStore';
import { HasAggregateId } from '../../domain/types/HasAggregateId';
import buildTestDataInFlatFormat from '../../test-data/buildTestDataInFlatFormat';
import { InMemoryDatabaseSnapshot } from '../../test-data/utilities';
import { DynamicDataTypeFinderService } from '../../validation';
import { ArangoConnectionProvider } from '../database/arango-connection.provider';
import { ArangoEdgeCollectionId } from '../database/collection-references/ArangoEdgeCollectionId';
import { ArangoDatabaseProvider } from '../database/database.provider';
import { HasArangoDocumentDirectionAttributes } from '../database/types/HasArangoDocumentDirectionAttributes';
import { ArangoDocumentForAggregateRoot } from '../database/utilities/mapEntityDTOToDatabaseDocument';
import { PersistenceModule } from '../persistence.module';
import generateDatabaseNameForTestSuite from './__tests__/generateDatabaseNameForTestSuite';
import TestRepositoryProvider from './__tests__/TestRepositoryProvider';
import { ArangoDataExporter } from './arango-data-exporter';
import path = require('path');

const testData = new DeluxeInMemoryStore(buildTestDataInFlatFormat());

const testDocs = [
    {
        _key: '123',
        name: 'bobby',
    },
    {
        _key: '124',
        name: 'billy',
    },
    {
        _key: '125',
        // ensure unicode chars come through
        name: 'ŵŝẑʔɨ’',
    },
];

const testEdges: HasArangoDocumentDirectionAttributes<
    ArangoDocumentForAggregateRoot<HasAggregateId>
>[] = [
    {
        _key: '333',
        _to: 'docs/123',
        _from: 'docs/124',
    },
    {
        _key: '334',
        _to: 'docs/123',
        _from: 'docs/123',
    },
];

const dir = `__cli-command-test-files__`;

const filename = 'data-dump__dump-file__.data.json';

const fullpath = path.join(dir, filename);

describe(`ArangoDataExporter`, () => {
    let app: INestApplication;

    let arangoDataExporter: ArangoDataExporter;

    let arangoDatabaseProvider: ArangoDatabaseProvider;

    const testEdgeCollectionName = 'edges';

    const testDocumentCollectionName = 'docs';

    const knownEdgeCollections = [...Object.values(ArangoEdgeCollectionId), testEdgeCollectionName];

    beforeAll(async () => {
        const testModule = await Test.createTestingModule({
            imports: [ConfigModule, PersistenceModule.forRootAsync()],
        })
            .overrideProvider(ConfigService)
            .useValue(
                buildMockConfigService(
                    {
                        ARANGO_DB_NAME: generateDatabaseNameForTestSuite(),
                    },
                    buildConfigFilePath(Environment.test)
                )
            )
            .compile();

        app = testModule.createNestApplication();

        await app.init();

        arangoDatabaseProvider = app.get(ArangoDatabaseProvider);

        arangoDataExporter = app.get(ArangoDataExporter);
    });

    beforeEach(async () => {
        await app
            .get(ArangoConnectionProvider)
            .createCollectionIfNotExists(testDocumentCollectionName);

        await arangoDatabaseProvider.getDatabaseForCollection(testDocumentCollectionName).clear();

        await arangoDatabaseProvider
            .getDatabaseForCollection(testDocumentCollectionName)
            .createMany(testDocs);

        await app.get(ArangoConnectionProvider).createCollectionIfNotExists(testEdgeCollectionName);

        await arangoDatabaseProvider.getDatabaseForCollection(testEdgeCollectionName).clear();

        await arangoDatabaseProvider
            .getDatabaseForCollection(testEdgeCollectionName)
            .createMany(testEdges);

        if (existsSync(fullpath)) {
            unlinkSync(fullpath);
        }

        const testRepositoryProvider = new TestRepositoryProvider(
            arangoDatabaseProvider,
            app.get(CoscradEventFactory),
            app.get(DynamicDataTypeFinderService)
        );

        await testRepositoryProvider.testSetup();

        // TODO add some views
        await testRepositoryProvider.addFullSnapshot(testData.fetchFullSnapshotInLegacyFormat());
    });

    describe('dumpSnapshot', () => {
        it(`should write all documents for all collections`, async () => {
            await arangoDataExporter.dumpSnapshot(dir, filename, knownEdgeCollections);

            const readResult = JSON.parse(
                readFileSync(fullpath, { encoding: 'utf-8' })
            ) as InMemoryDatabaseSnapshot;

            const foundDocs = readResult.document[testDocumentCollectionName]['documents'];

            expect(foundDocs).toHaveLength(testDocs.length);

            const foundEdges = readResult.edge[testEdgeCollectionName]['documents'];

            expect(foundEdges).toHaveLength(testEdges.length);

            expect(readResult).toMatchSnapshot();
        });
    });

    describe('restoreSnapshot', () => {
        describe(`when the checksums are valid`, () => {
            it(`should restore the snapshot`, async () => {
                const snapshot = await arangoDataExporter.fetchSnapshot(knownEdgeCollections);

                await arangoDataExporter.restoreFromSnapshot(snapshot);

                const foundTestDocs = await arangoDatabaseProvider
                    .getDatabaseForCollection(testDocumentCollectionName)
                    .fetchMany();

                expect(foundTestDocs).toHaveLength(testDocs.length);

                const foundEdges = await arangoDatabaseProvider
                    .getDatabaseForCollection(testEdgeCollectionName)
                    .fetchMany();

                expect(foundEdges).toHaveLength(testEdges.length);
            });
        });
    });
});
