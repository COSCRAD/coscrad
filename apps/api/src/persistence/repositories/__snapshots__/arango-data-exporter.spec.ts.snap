// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArangoDataExporter dumpSnapshot should write all documents for all collections 1`] = `
[
  {
    "_id": "docs/123",
    "_key": "123",
    "name": "bobby",
  },
  {
    "_id": "docs/124",
    "_key": "124",
    "name": "billy",
  },
  {
    "_id": "docs/125",
    "_key": "125",
    "name": "ŵŝẑʔɨ’",
  },
]
`;

exports[`ArangoDataExporter dumpSnapshot should write all documents for all collections 2`] = `
[
  {
    "_id": "edges/333",
    "_key": "333",
  },
  {
    "_id": "edges/334",
    "_key": "334",
  },
]
`;

exports[`ArangoDataExporter restoreSnapshot when attempting to restore to a non-empty database should throw 1`] = `[Error: Failed to import data to collection: photographs. Checksums do not match. Export: 9222859855719120303, post-import: 0]`;
